File: app/src/main/java/at/shockbytes/corey/fragment/RunningFragment.java
package at.shockbytes.corey.fragment;


import android.Manifest;
import android.content.IntentSender;
import android.graphics.Color;
import android.location.Location;
import android.os.Bundle;
import android.os.SystemClock;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v4.view.GestureDetectorCompat;
import android.view.GestureDetector;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.Chronometer;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.gms.common.api.ApiException;
import com.google.android.gms.common.api.ResolvableApiException;
import com.google.android.gms.location.LocationSettingsStatusCodes;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.Polyline;
import com.google.android.gms.maps.model.PolylineOptions;

import java.util.List;

import javax.inject.Inject;

import at.shockbytes.corey.R;
import at.shockbytes.corey.common.core.running.Run;
import at.shockbytes.corey.common.core.running.RunningManager;
import at.shockbytes.corey.common.core.running.location.LocationManager;
import at.shockbytes.corey.common.core.util.ResourceManager;
import at.shockbytes.corey.common.core.util.RunUtils;
import at.shockbytes.corey.core.CoreyApp;
import at.shockbytes.corey.core.RunningActivity;
import at.shockbytes.corey.storage.StorageManager;
import at.shockbytes.corey.util.AppParams;
import at.shockbytes.corey.util.ViewManager;
import butterknife.Bind;
import butterknife.ButterKnife;
import butterknife.OnClick;
import pub.devrel.easypermissions.AfterPermissionGranted;
import pub.devrel.easypermissions.EasyPermissions;


public class RunningFragment extends Fragment
        implements OnMapReadyCallback, LocationManager.OnLocationUpdateListener,
        GestureDetector.OnDoubleTapListener {

    private static final int REQ_CODE_PERM_LOCATION = 0x1245;
    private static final int REQUEST_CHECK_SETTINGS = 0x9874;

    public static RunningFragment newInstance() {
        RunningFragment fragment = new RunningFragment();
        Bundle args = new Bundle();
        fragment.setArguments(args);
        return fragment;
    }

    private GoogleMap map;
    private Polyline trackLine;
    private boolean isFirstLocation;

    private GestureDetectorCompat gestureDetector;

    @Inject
    protected LocationManager locationManager;

    @Inject
    protected RunningManager runningManager;

    @Inject
    protected StorageManager storageManager;

    @Bind(R.id.fragment_running_header)
    protected View headerView;

    @Bind(R.id.fragment_running_stop_help_view)
    protected View stopHelpView;

    @Bind(R.id.fragment_running_stop_help_imgview)
    protected View stopHelpImageView;

    @Bind(R.id.fragment_running_data_view)
    protected View headerDataView;

    @Bind(R.id.fragment_running_map_background)
    protected View mapBackgroundView;

    @Bind(R.id.fragment_running_btn_start)
    protected Button btnStart;

    @Bind(R.id.fragment_running_txt_time)
    protected Chronometer chronometer;

    @Bind(R.id.fragment_running_txt_distance)
    protected TextView txtDistance;

    @Bind(R.id.fragment_running_txt_current_pace)
    protected TextView txtCurrentPace;

    @Bind(R.id.fragment_running_txt_calories)
    protected TextView txtCalories;

    @Bind(R.id.fragment_running_txt_avg_pace)
    protected TextView txtAvgPace;

    public RunningFragment() {
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ((CoreyApp) getActivity().getApplication()).getAppComponent().inject(this);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_running, container, false);
        ButterKnife.bind(this, v);
        return v;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        setupTextViews();
        setupHeaderGestureRecognizer();
    }

    @Override
    public void onResume() {
        super.onResume();
        showMapFragment();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        ButterKnife.unbind(this);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        if (locationManager.isLocationUpdateRequested()) {
            locationManager.stop();
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
                                           @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        EasyPermissions.onRequestPermissionsResult(requestCode, permissions, grantResults, this);
    }

    @Override
    public void onMapReady(GoogleMap googleMap) {
        this.map = googleMap;
        setupMapAndLocationServices();
    }

    @SuppressWarnings({"MissingPermission"})
    @AfterPermissionGranted(REQ_CODE_PERM_LOCATION)
    private void setupMapAndLocationServices() {

        if (EasyPermissions.hasPermissions(getActivity(), Manifest.permission.ACCESS_COARSE_LOCATION,
                Manifest.permission.ACCESS_FINE_LOCATION)) {
            map.setMyLocationEnabled(true);
            locationManager.start(this);
        } else {
            EasyPermissions.requestPermissions(this, getString(R.string.perm_location_rationale),
                    REQ_CODE_PERM_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION,
                    Manifest.permission.ACCESS_FINE_LOCATION);
        }
    }

    private void setupTextViews() {
        chronometer.setText("00:00");
        txtDistance.setText("0.0 km");
        txtCurrentPace.setText("0:00\nmin/km");
        txtCalories.setText("0\nkcal");
        txtAvgPace.setText("0:00\nmin/km");
    }

    private void showMapFragment() {
        isFirstLocation = true;

        SupportMapFragment mapFragment = new SupportMapFragment();
        getFragmentManager().beginTransaction()
                .replace(R.id.fragment_running_map_container, mapFragment)
                .commit();
        mapFragment.getMapAsync(this);
    }

    private void setupHeaderGestureRecognizer() {

        gestureDetector = new GestureDetectorCompat(getContext(),
                new GestureDetector.OnGestureListener() {
                    @Override
                    public boolean onDown(MotionEvent motionEvent) {
                        return true;
                    }

                    @Override
                    public void onShowPress(MotionEvent motionEvent) {

                    }

                    @Override
                    public boolean onSingleTapUp(MotionEvent motionEvent) {
                        return true;
                    }

                    @Override
                    public boolean onScroll(MotionEvent motionEvent, MotionEvent motionEvent1, float v, float v1) {
                        return false;
                    }

                    @Override
                    public void onLongPress(MotionEvent motionEvent) {
                        showHelpHeader();
                    }

                    @Override
                    public boolean onFling(MotionEvent motionEvent, MotionEvent motionEvent1, float v, float v1) {
                        return false;
                    }
                });
        gestureDetector.setOnDoubleTapListener(this);

        headerView.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View view, MotionEvent motionEvent) {
                return gestureDetector.onTouchEvent(motionEvent);
            }
        });
    }

    private void showHelpHeader() {

        ((RunningActivity) getActivity()).animateToolbar();

        stopHelpView.animate()
                .alpha(1)
                .setStartDelay(0)
                .setDuration(400)
                .withEndAction(new Runnable() {
                    @Override
                    public void run() {
                        stopHelpView.animate()
                                .alpha(0)
                                .setStartDelay(AppParams.HELP_SHOW_DELAY)
                                .start();
                    }
                }).start();
        headerDataView.animate()
                .alpha(0.1f)
                .scaleY(0.8f)
                .scaleX(0.8f)
                .setStartDelay(0)
                .setDuration(400)
                .withEndAction(new Runnable() {
                    @Override
                    public void run() {
                        headerDataView.animate()
                                .alpha(1)
                                .scaleY(1)
                                .scaleX(1)
                                .setStartDelay(AppParams.HELP_SHOW_DELAY)
                                .start();
                    }
                }).start();

        ViewManager.animateDoubleTap(stopHelpImageView);
    }

    private void updateTrackOnMap(Location loc) {

        if (trackLine == null) {
            PolylineOptions lineOptions = new PolylineOptions()
                    .width(15)
                    .color(Color.parseColor("#03A9F4"));
            trackLine = map.addPolyline(lineOptions);
        }

        LatLng current = new LatLng(loc.getLatitude(), loc.getLongitude());
        List<LatLng> pointsSoFar = trackLine.getPoints();
        pointsSoFar.add(current);
        trackLine.setPoints(pointsSoFar);
    }

    private void updateViews(Run run) {

        long timeInMs = (SystemClock.elapsedRealtime() - chronometer.getBase());
        String averagePace = RunUtils.calculatePace(timeInMs, run.getDistance());
        String currentPace = runningManager.getCurrentPace();
        double weight = 80; // TODO Get latest weight from BodyManager
        int calories = RunUtils.calculateCaloriesBurned(run.getDistance(), weight);
        txtDistance.setText(ResourceManager.roundDoubleWithDigits(run.getDistance(), 2) + " km");
        txtAvgPace.setText(averagePace + "\nmin/km");

        txtCurrentPace.setText(currentPace + "\nmin/km");
        txtCalories.setText(calories + "\nkcal");
    }

    private void clearViews() {

        // Clear the text views
        setupTextViews();
        // Clear map
        if (trackLine != null) {
            trackLine.remove();
            trackLine = null;
        }
    }

    private void animateStartingViews(boolean animateOut) {

        int alpha = animateOut ? 0 : 1;

        // Animate button & transparent view with a fade out ;transition and hide it in the end
        btnStart.animate().alpha(alpha).setDuration(500);
        mapBackgroundView.animate().alpha(alpha).setDuration(500);
    }

    private void startRun() {

        clearViews();

        runningManager.startRunRecording();
        animateStartingViews(true);

        chronometer.setBase(SystemClock.elapsedRealtime());
        chronometer.start();

    }

    private void stopRun() {

        animateStartingViews(false);
        clearViews();

        chronometer.stop();
        long elapsedMillis = (SystemClock.elapsedRealtime() - chronometer.getBase());

        runningManager.stopRunRecord(elapsedMillis);

        Run run = runningManager.getFinishedRun();
        // TODO Enable StorageManager to handle runs storageManager.storeRun(run);
    }

    @OnClick(R.id.fragment_running_btn_start)
    protected void onClickButtonStart() {

        if (!runningManager.isRecording()) {
            startRun();
        }
    }

    @Override
    public void onConnected() {
    }

    @Override
    public void onDisconnected() {
    }

    @Override
    public void onError(Exception e) {

        int statusCode = ((ApiException) e).getStatusCode();
        switch (statusCode) {

            case LocationSettingsStatusCodes.RESOLUTION_REQUIRED:

                try {

                    ResolvableApiException rae = (ResolvableApiException) e;
                    rae.startResolutionForResult(getActivity(), REQUEST_CHECK_SETTINGS);

                } catch (IntentSender.SendIntentException sie) {
                    sie.printStackTrace();
                }

                break;

            case LocationSettingsStatusCodes.SETTINGS_CHANGE_UNAVAILABLE:

                Toast.makeText(getContext(), "Wrong location settings, fix in settings!",
                        Toast.LENGTH_LONG).show();
                break;

        }
    }

    @Override
    public void onLocationUpdate(Location location) {

        if (isFirstLocation) {
            isFirstLocation = false;
            map.moveCamera(CameraUpdateFactory
                    .newLatLngZoom(new LatLng(location.getLatitude(), location.getLongitude()), 16));
        } else {
            map.animateCamera(CameraUpdateFactory
                    .newLatLng(new LatLng(location.getLatitude(), location.getLongitude())));
        }

        if (runningManager.isRecording()) {
            Run run = runningManager.updateCurrentRun(location);
            updateViews(run);
            updateTrackOnMap(location);
        }
    }

    @Override
    public boolean onSingleTapConfirmed(MotionEvent motionEvent) {
        return true;
    }

    @Override
    public boolean onDoubleTap(MotionEvent motionEvent) {

        if (runningManager.isRecording()) {
            stopRun();
        } else {
            Snackbar.make(getView(), "Run hasn't started", Snackbar.LENGTH_SHORT).show();
        }
        return true;
    }

    @Override
    public boolean onDoubleTapEvent(MotionEvent motionEvent) {
        return true;
    }
}

File: app/src/main/java/at/shockbytes/corey/ui/fragment/tab/RunningFragment.kt
package at.shockbytes.corey.ui.fragment.tab

import android.Manifest
import android.annotation.SuppressLint
import android.graphics.Color
import android.os.Bundle
import android.os.SystemClock
import android.view.GestureDetector
import android.view.MotionEvent
import android.view.View
import androidx.core.view.GestureDetectorCompat
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import at.shockbytes.corey.R
import at.shockbytes.corey.common.addTo
import at.shockbytes.corey.common.core.location.CoreyLocation
import at.shockbytes.corey.common.core.running.Run
import at.shockbytes.corey.common.core.running.RunUpdate
import at.shockbytes.corey.common.core.running.RunViewAnimations
import at.shockbytes.corey.common.core.util.RunUtils
import at.shockbytes.corey.dagger.AppComponent
import at.shockbytes.corey.ui.viewmodel.RunningViewModel
import at.shockbytes.util.AppUtils
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.Polyline
import com.google.android.gms.maps.model.PolylineOptions
import kotlinx.android.synthetic.main.fragment_running.*
import pub.devrel.easypermissions.AfterPermissionGranted
import pub.devrel.easypermissions.EasyPermissions
import javax.inject.Inject
import kotlin.math.abs

class RunningFragment : TabBaseFragment<AppComponent>(),
    OnMapReadyCallback,
    GestureDetector.OnDoubleTapListener {

    private enum class HeaderState {
        NORMAL, TIME_UPFRONT, DISTANCE_UPFRONT
    }

    private enum class SwipeDirection {
        LEFT, RIGHT
    }

    private var map: GoogleMap? = null
    private var trackLine: Polyline? = null
    private var isFirstLocation: Boolean = false

    private var headerState: HeaderState = HeaderState.NORMAL

    private lateinit var gestureDetector: GestureDetectorCompat

    override val layoutId = R.layout.fragment_running

    override val castsActionBarShadow: Boolean = false

    override val snackBarBackgroundColorRes: Int = R.color.sb_background
    override val snackBarForegroundColorRes: Int = R.color.sb_background

    @Inject
    lateinit var vmFactory: ViewModelProvider.Factory

    private lateinit var viewModel: RunningViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        viewModel = ViewModelProviders.of(this, vmFactory)[RunningViewModel::class.java]
    }

    override fun bindViewModel() {

        viewModel.onRunStartedEvent()
            .subscribe {
                runStarted()
            }
            .addTo(compositeDisposable)

        viewModel.onRunStoppedEvent()
            .subscribe { run ->
                showFinishedRun(run)
            }
            .addTo(compositeDisposable)

        viewModel.onRunUpdate().observe(this, Observer { runUpdate ->
            updateRun(runUpdate)
        })
    }

    override fun injectToGraph(appComponent: AppComponent?) {
        appComponent?.inject(this)
    }

    override fun unbindViewModel() = Unit

    override fun setupViews() {
        clearViews()
        setupHeaderGestureRecognizer()

        fragment_running_btn_start.setOnClickListener {
            viewModel.startRun()
        }
    }

    override fun onResume() {
        super.onResume()
        showMapFragment()

        /* When running in background
        if (runningManager.isRecording) {
            setResetRunningViews(true)
            chronometer.base = runningManager.r.startTime
            chronometer.start()
        }
        */
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        EasyPermissions.onRequestPermissionsResult(requestCode, permissions, grantResults, this)
    }

    override fun onMapReady(googleMap: GoogleMap) {
        this.map = googleMap
        setupMapAndLocationServices()
    }

    override fun onSingleTapConfirmed(motionEvent: MotionEvent): Boolean {
        return true
    }

    override fun onDoubleTap(motionEvent: MotionEvent): Boolean {
        stopRun()
        return true
    }

    override fun onDoubleTapEvent(motionEvent: MotionEvent): Boolean {
        return true
    }

    @SuppressLint("MissingPermission")
    @AfterPermissionGranted(REQ_CODE_PERM_LOCATION)
    private fun setupMapAndLocationServices() {

        if (EasyPermissions.hasPermissions(requireActivity(), Manifest.permission.ACCESS_COARSE_LOCATION,
                Manifest.permission.ACCESS_FINE_LOCATION)) {
            map?.isMyLocationEnabled = true
        } else {
            EasyPermissions.requestPermissions(this, getString(R.string.perm_location_rationale),
                REQ_CODE_PERM_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION,
                Manifest.permission.ACCESS_FINE_LOCATION)
        }
    }

    // -------------------------- Setup --------------------------

    private fun setupHeaderGestureRecognizer() {

        gestureDetector = GestureDetectorCompat(context,
            object : GestureDetector.OnGestureListener {

                override fun onDown(motionEvent: MotionEvent): Boolean {
                    return true
                }

                override fun onShowPress(motionEvent: MotionEvent) {}

                override fun onSingleTapUp(motionEvent: MotionEvent): Boolean {
                    return true
                }

                override fun onScroll(motionEvent: MotionEvent, motionEvent1: MotionEvent, v: Float, v1: Float): Boolean {
                    return false
                }

                override fun onLongPress(motionEvent: MotionEvent) {
                    showHelpHeader()
                }

                override fun onFling(downEvent: MotionEvent, upEvent: MotionEvent, vx: Float, vy: Float): Boolean {

                    val directionX = downEvent.x - upEvent.x
                    if (abs(directionX) > 200) {
                        val direction = if (directionX > 0)
                            SwipeDirection.LEFT
                        else
                            SwipeDirection.RIGHT
                        animateTimeDistanceHeader(direction)
                    }
                    return true
                }
            })
        gestureDetector.setOnDoubleTapListener(this)

        fragment_running_header.setOnTouchListener { _, motionEvent ->
            gestureDetector.onTouchEvent(motionEvent)
        }
    }

    private fun showMapFragment() {
        isFirstLocation = true

        val mapFragment = SupportMapFragment()
        childFragmentManager.beginTransaction()
            .replace(R.id.fragment_running_map_container, mapFragment)
            .commit()
        mapFragment.getMapAsync(this)
    }

    private fun showHelpHeader() {

        fragment_running_stop_help_view.animate()
            .alpha(1f)
            .setStartDelay(0)
            .setDuration(400)
            .withEndAction {
                fragment_running_stop_help_view.animate()
                    .alpha(0f)
                    .setStartDelay(2000L)
                    .start()
            }.start()
        fragment_running_data_view.animate()
            .alpha(0.1f)
            .scaleY(0.8f)
            .scaleX(0.8f)
            .setStartDelay(0)
            .setDuration(400)
            .withEndAction {
                fragment_running_data_view.animate()
                    .alpha(1f)
                    .scaleY(1f)
                    .scaleX(1f)
                    .setStartDelay(2000L)
                    .start()
            }.start()

        RunViewAnimations.animateDoubleTap(fragment_running_stop_help_imgview)
    }

    private fun showFinishedRun(run: Run) {
        fragment_running_txt_time.stop()
        setResetRunningViews(false)
        showDetailFragment(run)
    }

    private fun showDetailFragment(run: Run) {
        // Show detail fragment, old implementation wasn't nice anyway
    }

    private fun animateStartingViews(animateOut: Boolean) {

        if (animateOut) {

            fragment_running_btn_start
                .animate()
                .alpha(0f)
                .setDuration(500)
                .withEndAction {
                    fragment_running_btn_start.visibility = View.GONE
                }
                .start()

            fragment_running_map_background
                .animate()
                .alpha(0f)
                .setDuration(500)
                .withEndAction {
                    fragment_running_btn_start.visibility = View.GONE
                }
                .start()
        } else {

            fragment_running_btn_start
                .animate()
                .alpha(1f)
                .setDuration(500)
                .withStartAction() {
                    fragment_running_btn_start.visibility = View.VISIBLE
                }
                .start()

            fragment_running_map_background
                .animate()
                .alpha(1f)
                .setDuration(500)
                .withStartAction() {
                    fragment_running_btn_start.visibility = View.VISIBLE
                }
                .start()
        }
    }

    private fun animateTimeDistanceHeader(direction: SwipeDirection) {

        val x = (fragment_running_data_view.width / 2 -
            fragment_running_txt_distance.width / 2 -
            fragment_running_txt_time.x.toInt())

        when (headerState) {

            HeaderState.NORMAL ->

                headerState = if (direction == SwipeDirection.LEFT) {
                    fragment_running_txt_distance
                        .animate()
                        .translationX((-x).toFloat())
                        .start()
                    fragment_running_txt_time
                        .animate()
                        .scaleX(0.5f)
                        .scaleY(0.5f)
                        .alpha(0.5f)
                        .start()

                    HeaderState.DISTANCE_UPFRONT
                } else {
                    fragment_running_txt_time
                        .animate()
                        .translationX(x.toFloat())
                        .start()
                    fragment_running_txt_distance
                        .animate()
                        .scaleX(0.5f)
                        .scaleY(0.5f)
                        .alpha(0.5f)
                        .start()

                    HeaderState.TIME_UPFRONT
                }

            HeaderState.TIME_UPFRONT ->

                if (direction == SwipeDirection.LEFT) {
                    fragment_running_txt_time
                        .animate()
                        .translationX(0f)
                        .start()
                    fragment_running_txt_distance
                        .animate()
                        .scaleX(1f)
                        .scaleY(1f)
                        .alpha(1f)
                        .start()

                    headerState = HeaderState.NORMAL
                }

            HeaderState.DISTANCE_UPFRONT ->

                if (direction == SwipeDirection.RIGHT) {
                    fragment_running_txt_distance
                        .animate()
                        .translationX(0f)
                        .start()
                    fragment_running_txt_time
                        .animate()
                        .scaleX(1f)
                        .scaleY(1f)
                        .alpha(1f)
                        .start()

                    headerState = HeaderState.NORMAL
                }
        }
    }

    private fun runStarted() {

        val startMillis = SystemClock.elapsedRealtime()
        fragment_running_txt_time.base = startMillis
        fragment_running_txt_time.start()
        setResetRunningViews(true)
    }

    private fun stopRun() {
        viewModel.stopRun()
    }

    private fun updateRun(runUpdate: RunUpdate) {

        val location = runUpdate.currentLocation
        if (isFirstLocation) {
            isFirstLocation = false
            map?.moveCamera(CameraUpdateFactory
                .newLatLngZoom(LatLng(location.lat, location.lng), 16f))
        } else {
            map?.animateCamera(CameraUpdateFactory
                .newLatLng(LatLng(location.lat, location.lng)))
        }

        updateViews(runUpdate)
        setTrackOnMap(runUpdate.locations)
    }

    @SuppressLint("SetTextI18n")
    private fun updateViews(update: RunUpdate) {

        val timeInMs = SystemClock.elapsedRealtime() - fragment_running_txt_time.base
        val averagePace = RunUtils.calculatePace(timeInMs, update.distance)
        val calories = RunUtils.calculateCaloriesBurned(timeInMs.toDouble(), update.userWeight)
        fragment_running_txt_distance.text = "${AppUtils.roundDouble(update.distance, 2)}km"
        fragment_running_txt_avg_pace.text = "$averagePace\nmin/km"
        fragment_running_txt_current_pace.text = update.currentPace + "\nmin/km"
        fragment_running_txt_calories.text = "$calories\nkcal"
    }

    private fun setTrackOnMap(runPoints: List<CoreyLocation>) {

        if (trackLine == null) {
            val lineOptions = PolylineOptions()
                .width(15f)
                .color(Color.parseColor("#03A9F4"))
            trackLine = map?.addPolyline(lineOptions)
        }
        trackLine?.points = runPoints.map { LatLng(it.lat, it.lng) }
    }

    @SuppressLint("SetTextI18n")
    private fun clearViews() {

        // Clear the text views
        fragment_running_txt_time.text = "00:00"
        fragment_running_txt_distance.text = "0.0 km"
        fragment_running_txt_current_pace.text = "0:00\nmin/km"
        fragment_running_txt_calories.text = "0\nkcal"
        fragment_running_txt_avg_pace.text = "0:00\nmin/km"

        // Clear map
        if (trackLine != null) {
            trackLine?.remove()
            trackLine = null
        }
    }

    private fun setResetRunningViews(isSetup: Boolean) {
        clearViews()
        animateStartingViews(isSetup)
    }

    companion object {

        private const val REQ_CODE_PERM_LOCATION = 0x7916

        fun newInstance(): RunningFragment {
            val fragment = RunningFragment()
            val args = Bundle()
            fragment.arguments = args
            return fragment
        }
    }
}